### ISSUE TITLE
Cloudvars in old projects (i.e. saved with Scratch 2) are 'write-only'

### ISSUE THREAD

**Issue by TheLogFather** [LLK/scratch-gui#4723]
I've been holding off reporting this in the hope it might've been a temporary issue that got corrected as more development was done on cloud.
Also, it's not that clear to me which section cloud issues (often others too) should go into...

### Expected Behavior

All pre-Scratch 3 projects still work correctly with cloud.

### Actual Behavior

Old cloud-based projects (i.e. not re-saved since Scratch 3) cannot 'collect' values from cloud, so only see the value of a cloudvar at the point the project was last saved.

However, if an old project changes a cloudvar, it *does* make it through to the cloud server (and, of course, the local project sees the change, since it sees its own local copy – but it hasn't got that new value from cloud, and nobody else can see the new value because their own instances of the project don't see values in cloud, only values saved with the project).

In other words, cloudvars for old (.sb2?) projects are effectively 'write-only'.

### Steps to Reproduce

1. Go to this old (unshared) cloud project: https://scratch.mit.edu/projects/135513871/
2. See inside, note that the cloudvar has value "?". This value is clearly not numerical, so has not come from cloud –it was the value saved with the project. (The value in cloud is currently one, recently set by me, as you will soon see below from the cloud log page.)
3. Go to the 'variables' section of the blocks palette and click "change by 1" –now it will show value one in the project (since the value of "?" was considered to be zero).
4. Go to the cloud monitor page and note that cloud did see your change.
5. Reload the project, and see inside that the cloudvar still loads with value "?".

Can also check out other old cloud-based projects, and note that the values you see for the cloudvars often do not correspond with those in the cloud (but correspond with the values saved with the project).

### Operating System and Browser

Probably N/A (checked both Safari & Firefox on macOS).

**Comment by paulkaplan** [comment]
Thanks for detailed report. Are you able to reproduce this problem with a variable that was a number instead of one that was a string? It could be an issue with the fact that somehow that cloud variable got a string value.

**Comment by TheLogFather** [comment]
No, the non-numeric value makes no difference – the project I chose above was simply some old unshared junk I had lying around that had the appropriate characteristics to demo the problem.

However, it's the same for *all* of my old cloud projects (on both accounts) – none of them see values from the cloud, only the values that were last saved with the project.

----

Here's one where I regularly get comments saying it doesn't work (and I regularly reply to explain why): https://scratch.mit.edu/projects/18593258

Note that the value in the cloud is 1, but everyone always just sees the value zero (which is what was saved with the project).

Also note that the cloud value has been changed multiple times (a couple of months ago) from within the project – not only by me, but by someone else who was (ab)using the cloudvar slider loophole to change it while in the editor (which is possible because the cloudvar monitor is showing).

None of those changes (to non-zero values) are seen on reloading the project – only ever see zero.

----

This is a multiplayer demo project I wrote a while back: https://scratch.mit.edu/projects/205282370/

If you run the project, and 'join' for a while, and then look at the cloud log, it's clear that everything has got through to cloud just fine.

But if you close the tab while in the middle of using it, so your cloudvar never gets chance to reset to zero, then you can see it has a non-zero value in cloud, but reloading still gives zero (the value when project was last saved).

**Comment by TheLogFather** [comment]
UPDATE:

I've tested the above multiplayer project again just now, and it appears to be working ok. –Not sure when that changed, as it's a while since I tested it properly.

However, what I can still say is that the cloudvars do not get their values from cloud after first loading the project. A new value only come from cloud if it is *subsequently* changed after loading.

This is definitely a big improvement, but it does mean that the first project I referenced in the comment just above (https://scratch.mit.edu/projects/18593258) is still not working correctly (because that doesn't actually change any cloudvars, only compares value in the cloud against value loaded with the project).

**Comment by ArnoHue** [comment]
Thanks a lot for reporting this! I ran into the same behavior, namely that on project start the user sees cloud variable state only from when the project was saved. I thought that might be a temporary problem, but it wasn't, and I had to rewrite my online chess project (https://scratch.mit.edu/projects/209130998/) to wait for a while after project start, until I could be sure to have received the current game state or that no one else would be online (before actually updating any state).

My impression is that this problem just started some weeks ago (and that it worked when Scratch 3.0 came out).

On the other hand, this little pacman project (https://scratch.mit.edu/projects/97137611/) still seems to update highscores correctly. Both projects were created with Scatch 2.0.
